src/java/org/apache/cassandra/service/reads/DigestResolver.java

```diff
@@ -109,7 +109,8 @@ public boolean responsesMatch()
         // validate digests against each other; return false immediately on mismatch.
         ByteBuffer digest = null;
         Collection<Message<ReadResponse>> snapshot = responses.snapshot();
-        if (snapshot.size() <= 1)
+        assert snapshot.size() > 0;
```

Comment:
`! isEmpty`, that "can" be cheaper for collections than size (even though most implement it as a size check).

Reply:
Also, good to add a assert message saying why this failed.

Reply:
I'm don't want to be argumentative here, but of 1200 asserts in the codebase, 1027 do not have messages. To my way of thinking this is just codifying an assumption. Why it failed is because our code called this method under unexpected conditions, i.e. there are no responses. A message might not add anything to the assertion error. Do you think maybe it would be worth adding language like "This is unexpected, and likely indicates a bug. Please file a bug report at https://issues.apache.org/jira/projects/CASSANDRA/summary"?

Reply:
> but of 1200 asserts in the codebase, 1027 do not have messages

Yes, and when I hit them I end up fixing it as "there is an error, read the code" is really hard to deal with... harder when working with members of the community that run forks, so line numbers don't match.

> "This is unexpected, and likely indicates a bug. Please file a bug report at https://issues.apache.org/jira/projects/CASSANDRA/summary"

Same issue as the above, "there is a bug" isn't informative, but something like `no responses found` or `responses was empty` is enough.

> A message might not add anything to the assertion error

I strongly disagree here as I find its been incredibly painful debugging external users issues and have spent days going down rabbit holes.  These asserts that should never happen have a way of popping up in the wild, so better to be more clear to the person debugging what's going on.

```
assert snapshot.size() > 0 : "no responses found";
```

this is enough for me.

Reply:
> harder when working with members of the community that run forks, so line numbers don't match

This is the thing that usually tips the scales for me. I know we've been bad about this in the past, but especially given we don't ever seem to actually `-da` anything, even a short message is better than just a line number and trace.

Reply:
I don't intend to argue this anymore. I was just letting the line site while we try to come to consensus in the thread over here:
https://github.com/aholmberg/cassandra/pull/17#discussion_r525335373

Anybody have an opinion on this logic or the proposed alternative?